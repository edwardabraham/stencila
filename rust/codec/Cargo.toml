[package]
name = "codec"
description = "Defines a specification and trait for codecs"
version = "0.0.0"
edition = "2021"

[dependencies]
common = { path = "../common" }
stencila-schema = { path = "../schema" }
utils = { path = "../utils" }
