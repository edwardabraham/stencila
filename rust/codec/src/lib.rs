use std::{fs::create_dir_all, path::Path};

use common::{
    async_trait::async_trait,
    eyre::{bail, Result},
    serde::{Deserialize, Serialize},
    tokio::{
        fs::File,
        io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt, BufReader, BufWriter},
    },
};
use stencila_schema::Node;
use utils::vec_string;

// Re-export for the convenience of crates that implement `CodecTrait`
pub use common;
pub use stencila_schema;
pub use utils;

/// A specification for codecs
///
/// All codecs, including those implemented in plugins, should provide this
/// specification. Rust implementations return a `Codec` instance from the
/// `spec` function of `CodecTrait`. Plugins provide a JSON or YAML serialization
/// as part of their manifest.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(default, rename_all = "camelCase", crate = "common::serde")]
pub struct Codec {
    /// A list of format names (or aliases) that the codec can decode from
    /// or encode to
    pub formats: Vec<String>,

    /// Whether the codec supports decoding from string content
    pub from_string: bool,

    /// Whether the codec supports decoding from a file system path
    pub from_path: bool,

    /// Whether the codec supports encoding to string content
    pub to_string: bool,

    /// Whether the codec supports encoding to a file system path
    pub to_path: bool,

    /// A list of root node types that the codec can encode / decode
    ///
    /// Most codecs usually only handle one root type e.g. `Article`.
    /// Used to provide a list of formats to the user that support
    /// the current document type.
    pub root_types: Vec<String>,

    /// A list of node types that the codec does not support
    ///
    /// Used to provide warnings to the user on potential loss of content
    /// when encoding using the codec.
    pub unsupported_types: Vec<String>,

    /// A list of node properties that the codec does not support
    ///
    /// The format for these strings is `<type>.<property>` e.g. `Article.funders`
    /// Used to provide warnings to the user on potential loss of content
    /// when encoding using the codec.
    pub unsupported_properties: Vec<String>,

    /// The status of the codec e.g. `alpha`, `beta`, `stable`
    pub status: String,
}

impl Default for Codec {
    fn default() -> Self {
        Self {
            formats: vec![],
            from_string: true,
            from_path: true,
            to_string: true,
            to_path: true,
            root_types: vec_string!["Article"],
            unsupported_types: vec![],
            unsupported_properties: vec![],
            status: "alpha".to_string(),
        }
    }
}

/// A trait for codecs
///
/// This trait can be used by Rust implementations of codecs, allowing them to
/// be compiled into the Stencila binaries.
///
/// It defines similar functions to `serde_json` (and other `serde_` crates) for
/// converting nodes to/from strings, files, readers etc.
#[async_trait]
pub trait CodecTrait {
    /// Get the [`Codec`] specification for this implementation
    fn spec() -> Codec;

    /// Decode a document node from a string
    fn from_str(_str: &str, _options: Option<DecodeOptions>) -> Result<Node> {
        bail!("Decoding from string is not implemented for this format")
    }

    /// Decode a document node from a string asynchronously
    async fn from_str_async(str: &str, options: Option<DecodeOptions>) -> Result<Node> {
        Self::from_str(str, options)
    }

    /// Decode a document node from a `BufReader`
    async fn from_buffer<T: AsyncRead>(
        reader: &mut BufReader<T>,
        options: Option<DecodeOptions>,
    ) -> Result<Node>
    where
        T: Send + Sync + Unpin,
    {
        let mut content = String::new();
        reader.read_to_string(&mut content).await?;
        Self::from_str_async(&content, options).await
    }

    /// Decode a document node from a file
    ///
    /// This function reads the file as a string and passes that on to `from_str`
    /// for decoding. If working with binary formats, you should override this function
    /// to read the file as bytes instead.
    async fn from_file(file: &mut File, options: Option<DecodeOptions>) -> Result<Node> {
        let mut content = String::new();
        file.read_to_string(&mut content).await?;
        Self::from_str_async(&content, options).await
    }

    /// Decode a document node from a file system path
    async fn from_path(path: &Path, options: Option<DecodeOptions>) -> Result<Node> {
        let mut file = File::open(path).await?;
        Self::from_file(&mut file, options).await
    }

    /// Encode a document node to a string
    fn to_string(_node: &Node, _options: Option<EncodeOptions>) -> Result<String> {
        bail!("Encoding to string is not implemented for this format")
    }

    /// Encode a document node to a string asynchronously
    async fn to_string_async(node: &Node, options: Option<EncodeOptions>) -> Result<String> {
        Self::to_string(node, options)
    }

    /// Encode a document node to a `BufWriter`
    async fn to_buffer<T: AsyncWrite>(
        node: &Node,
        writer: &mut BufWriter<T>,
        options: Option<EncodeOptions>,
    ) -> Result<()>
    where
        T: Send + Sync + Unpin,
    {
        let content = Self::to_string_async(node, options).await?;
        writer.write_all(content.as_bytes()).await?;
        Ok(())
    }

    /// Encode a document node to a file
    async fn to_file(node: &Node, file: &mut File, options: Option<EncodeOptions>) -> Result<()> {
        let content = Self::to_string_async(node, options).await?;
        file.write_all(content.as_bytes()).await?;
        Ok(())
    }

    /// Encode a document node to a file system path
    async fn to_path(node: &Node, path: &Path, options: Option<EncodeOptions>) -> Result<()> {
        if let Some(parent) = path.parent() {
            create_dir_all(parent)?
        }
        let mut file = File::create(path).await?;
        Self::to_file(node, &mut file, options).await
    }
}

/// Decoding options
///
/// Decoding functions (including those in plugins) are encouraged to respect these options
/// but are not required to.
#[derive(Clone, Default)]
pub struct DecodeOptions {
    /// The format of the input to be decoded
    ///
    /// Most codecs only decode one format. However, for those that handle multiple
    /// format it may be necessary to specify this option.
    pub format: Option<String>,
}

/// Encoding options
///
/// Encoding functions (including those in plugins) are encouraged to respect these options
/// but are not required to. Indeed, some options do not apply for some formats.
/// For example, a PDF is always `standalone` (so if that option is set to `false`, it will be ignored).
/// Futhermore, some combinations of options are ineffectual e.g. a `theme` when `standalone: false`
#[derive(Clone)]
pub struct EncodeOptions {
    /// Whether to encode in compact form
    ///
    /// Some formats (e.g HTML and JSON) can be encoded in either compact
    /// or "pretty-printed" (e.g. indented) forms.
    pub compact: bool,

    /// Whether to ensure that the encoded document is standalone
    ///
    /// Some formats (e.g. Markdown, DOCX) are always standalone.
    /// Others can be fragments, or standalone documents (e.g HTML).
    pub standalone: bool,

    /// Whether to bundle local media files into the encoded document
    ///
    /// Some formats (e.g. DOCX, PDF) always bundle. For HTML, bundling means
    /// including media as data URIs rather than links to files.
    pub bundle: bool,

    /// Whether to use Stencila Cloud as a remote storage for document nodes
    /// that can not be wholly represented within the format
    ///
    /// Some formats (e.g. Google Docs) are not able to fully represent some types
    /// of nodes (e.g. CodeChunks). In these cases we store the node on Stencila Cloud
    /// with a unique id and key and insert a link to it from within the format.
    pub rpng_content: bool,
    pub rpng_link: bool,

    /// The theme to apply to the encoded document
    ///
    /// Only applies to standalone HTML and formats derived from it (e.g. PDF, PNG).
    pub theme: Option<String>,

    /// Whether Javascript for Web Components should be loaded or not
    ///
    /// Only applies to standalone HTML and formats derived from it (e.g. PDF, PNG).
    pub components: bool,

    /// The format to encode to
    ///
    /// Most codecs only encode to one format. However, for those that handle multiple
    /// formats it may be necessary to specify this option.
    pub format: Option<String>,
}

impl Default for EncodeOptions {
    fn default() -> Self {
        Self {
            compact: true,
            standalone: false,
            bundle: false,
            rpng_content: false,
            rpng_link: false,
            theme: None,
            components: true,
            format: None,
        }
    }
}
