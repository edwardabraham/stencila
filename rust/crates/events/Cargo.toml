[package]
name = "events"
version = "0.0.0"
edition = "2021"

[dependencies]
serde = { version = "1.0.130", features = ["derive"] }
