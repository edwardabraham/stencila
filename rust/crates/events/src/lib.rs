use serde::{Deserialize, Serialize};

/// An event updating progress of some task
#[derive(Default, Debug, Deserialize, Serialize)]
pub struct ProgressEvent {
    /// The id of the task that this progress event relates to
    pub id: Option<String>,

    /// The id of the parent task (if any)
    pub parent: Option<String>,

    /// The event message
    pub message: Option<String>,

    /// The current value
    pub current: Option<i64>,

    /// The expected value when complete
    pub expected: Option<i64>,

    // Whether or not the task is complete
    pub complete: bool,
}
