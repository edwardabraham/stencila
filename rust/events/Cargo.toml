[package]
name = "events"
version = "0.0.0"
edition = "2021"

[dependencies]
eyre = "0.6.5"
once_cell = "1.9.0"
serde = { version = "1.0.130", features = ["derive"] }
serde_json = "1.0.75"
tokio = { version = "1.13.0", features = ["sync", "signal"] }
tracing = "0.1.29"
uuids = { path = "../uuids", version = "0.0.0" }
