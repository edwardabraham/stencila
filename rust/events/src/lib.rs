use eyre::{bail, Result};
use once_cell::sync::Lazy;
use serde::{Deserialize, Serialize};
use std::{
    collections::HashMap,
    sync::{Mutex, MutexGuard},
};
use tokio::sync::mpsc;
use uuids::uuid_family;

/// An event updating progress of some task
#[derive(Default, Debug, Deserialize, Serialize)]
pub struct ProgressEvent {
    /// The id of the task that this progress event relates to
    pub id: Option<String>,

    /// The id of the parent task (if any)
    pub parent: Option<String>,

    /// The event message
    pub message: Option<String>,

    /// The current value
    pub current: Option<i64>,

    /// The expected value when complete
    pub expected: Option<i64>,

    // Whether or not the task is complete
    pub complete: bool,
}

pub type Message = (String, serde_json::Value);

uuid_family!(SubscriptionId, "su");

pub enum Subscriber {
    Function(fn(topic: String, event: serde_json::Value) -> ()),
    Sender(mpsc::UnboundedSender<Message>),
}

struct Subscription {
    /// The topic that is subscribed to
    topic: String,

    /// The subscriber that is sent an event
    subscriber: Subscriber,
}

/// The glocal subscriptions store
static SUBSCRIPTIONS: Lazy<Mutex<HashMap<SubscriptionId, Subscription>>> =
    Lazy::new(|| Mutex::new(HashMap::new()));

/// Obtain the subscriptions store
fn obtain() -> Result<MutexGuard<'static, HashMap<SubscriptionId, Subscription>>> {
    // Use `lock`, not `try_lock`, which means this thread may block waiting
    // a little for the subscriptions to become available.
    match SUBSCRIPTIONS.lock() {
        Ok(guard) => Ok(guard),
        Err(error) => bail!("While attempting to obtain subscriptions: {}", error),
    }
}

/// Subscribe to a topic
pub fn subscribe(topic: &str, subscriber: Subscriber) -> Result<SubscriptionId> {
    tracing::debug!("Subscribing to topic `{}`", topic);

    match obtain() {
        Ok(mut subscriptions) => {
            let id = SubscriptionId::new();
            subscriptions.insert(
                id.clone(),
                Subscription {
                    topic: topic.to_string(),
                    subscriber,
                },
            );
            Ok(id)
        }
        Err(error) => {
            bail!("Unable to subscribe: {}", error.to_string())
        }
    }
}

/// Unsubscribe
pub fn unsubscribe(subscription_id: SubscriptionId) -> Result<()> {
    tracing::debug!("Unsubscribing subscription `{}`", subscription_id);

    match obtain() {
        Ok(mut subscriptions) => {
            subscriptions.remove(&subscription_id);
            Ok(())
        }
        Err(error) => {
            bail!("Unable to unsubscribe: {}", error.to_string())
        }
    }
}

/// Publish an event for a topic
///
/// Publishing an event should be treated as 'fire-and-forget'.
/// This function does not return an `Err` if it fails but will
/// log an error (if not already attempting to publish to logging channel).
pub fn publish<Event>(topic: &str, event: Event)
where
    Event: Serialize,
{
    if topic != "logging" {
        tracing::debug!("Publishing event for topic `{}`", topic);
    }

    match obtain() {
        Ok(subscriptions) => {
            for subscription in subscriptions.values() {
                if subscription.topic == "*"
                    || subscription.topic == topic
                    || topic.starts_with(&subscription.topic)
                {
                    let value = serde_json::to_value(&event).unwrap_or(serde_json::Value::Null);
                    match &subscription.subscriber {
                        Subscriber::Function(function) => {
                            function(topic.into(), value);
                        }
                        Subscriber::Sender(sender) => {
                            if let Err(error) = sender.send((topic.into(), value)) {
                                tracing::error!("Error sending event: {}", error);
                            }
                        }
                    }
                }
            }
        }
        Err(error) => {
            // Do not log error if the topic is logging since that could lead to recursion
            if topic != "logging" {
                tracing::error!("Unable to publish event: {}", error)
            }
        }
    }
}
