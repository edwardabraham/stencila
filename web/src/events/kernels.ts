import { FileFormatUtils } from '@stencila/components'

type DiscoverExecutableLanguagesEvent = CustomEvent<{
  kernels: string[]
  executableLanguages: FileFormatUtils.FileFormatMap
}>

const kernelsToFileFormatMap = (
  kernels: string[]
): FileFormatUtils.FileFormatMap =>
  kernels.reduce((formats: FileFormatUtils.FileFormatMap, kernelName) => {
    const foundFormat = FileFormatUtils.lookupFormat(kernelName)
    return { ...formats, [foundFormat.name]: foundFormat }
  }, {})

/**
 * Emit a CustomEvent with the name of`'stencila-discover-kernels'` detailing
 * available executable kernels. Custom WebComponents listen to this event and
 * update the list of executable languages in the language selector fields.
 *
 * We also update the global `window.stencilaWebClient.executableLanguages` field
 * so that any components added after the event has fired can be instantiated with
 * an up to date list of languages.
 */
export const onDiscoverExecutableLanguages = (
  kernels: string[]
): {
  kernels: string[]
  executableLanguages: FileFormatUtils.FileFormatMap
} => {
  const executableLanguages = kernelsToFileFormatMap(kernels)
  const event: DiscoverExecutableLanguagesEvent = new CustomEvent(
    'stencila-discover-kernels',
    {
      detail: {
        kernels,
        executableLanguages,
      },
    }
  )

  if (window.stencilaWebClient) {
    window.stencilaWebClient.executableLanguages = executableLanguages
  }

  window.dispatchEvent(event)

  return {
    kernels,
    executableLanguages,
  }
}
