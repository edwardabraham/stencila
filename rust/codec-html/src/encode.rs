use codec::{eyre::Result, EncodeOptions};
use html_escape::{encode_double_quoted_attribute, encode_safe};
use inflector::cases::camelcase::to_camel_case;
use std::any::type_name;
use stencila_schema::*;

/// Encode a `Node` to a HTML document
pub fn encode(node: &Node, options: Option<EncodeOptions>) -> Result<String> {
    let html = encode_root(node, options.clone());

    let EncodeOptions {
        theme, standalone, ..
    } = options.unwrap_or_default();

    let html = if standalone {
        wrap_standalone("", &theme, &html)
    } else {
        html
    };

    Ok(html)
}

/// Generate the HTML fragment for a root node
///
/// This function is used when translating a `Operation` (where any value of
/// the operation is a `Node` and the operation is applied to a `Node`) to a `DomOperation`
/// (where any value is either a HTML or JSON string and the operation is applied to a browser DOM).
pub fn encode_root(node: &Node, options: Option<EncodeOptions>) -> String {
    let EncodeOptions {
        bundle, compact, ..
    } = options.unwrap_or_default();

    let context = EncodeContext { root: node, bundle };
    let html = node.to_html(&context);

    // Add the `data-itemscope="root"` attribute.
    // This is currently used in `themes` (for CSS scope) and in `web` (for address resolution).
    // This is a bit hacky and there may be a better approach. Or we may find
    // a way of avoid this entirely.
    let html = html.replacen(" ", " data-itemscope=\"root\" ", 1);

    if compact {
        html
    } else {
        indent(&html)
    }
}

/// Indent generated HTML
///
/// Originally based on https://gist.github.com/lwilli/14fb3178bd9adac3a64edfbc11f42e0d
fn indent(html: &str) -> String {
    use quick_xml::events::Event;
    use quick_xml::{Reader, Writer};

    let mut buf = Vec::new();

    let mut reader = Reader::from_str(html);
    reader.trim_text(true);

    let mut writer = Writer::new_with_indent(Vec::new(), b' ', 2);

    loop {
        let ev = reader.read_event(&mut buf);

        match ev {
            Ok(Event::Eof) => break,
            Ok(event) => writer.write_event(event),
            Err(e) => panic!("Error at position {}: {:?}", reader.buffer_position(), e),
        }
        .expect("Failed to parse XML");

        buf.clear();
    }

    std::str::from_utf8(&*writer.into_inner())
        .expect("Failed to convert a slice of bytes to a string slice")
        .to_string()
}

/// Wrap generated HTML so that it is standalone
pub fn wrap_standalone(title: &str, theme: &str, html: &str) -> String {
    let title = if title.is_empty() { "Untitled" } else { &title };
    let theme = if theme.is_empty() { "stencila" } else { &theme };

    format!(
        r#"<!DOCTYPE html>
<html lang="en">
    <head>
        <title>{title}</title>
        <meta charset="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <link
            href="https://unpkg.com/@stencila/thema/dist/themes/{theme}/styles.css"
            rel="stylesheet">
        <script
            src="https://unpkg.com/@stencila/components/dist/stencila-components/stencila-components.esm.js"
            type="module"></script>
        <script
            src="https://unpkg.com/@stencila/components/dist/stencila-components/stencila-components.js"
            nomodule=""></script>
        <style>
            .error {{
                font-family: mono;
                color: #9e0000;
                background: #ffd9d9;
            }}
            .todo {{
                font-family: mono;
                color: #9e9b00;
                background: #faf9de;
            }}
            .unsupported {{
                font-family: mono;
                color: #777;
                background: #eee;
            }}
        </style>
    </head>
    <body>
        {html}
    </body>
</html>"#,
        title = title,
        theme = theme,
        html = html
    )
}

/// The encoding context.
///
/// Used by child nodes to retrieve necessary information about the
/// parent nodes when rendering themselves.
pub struct EncodeContext<'a> {
    /// The root node being encoded
    root: &'a Node,

    /// Whether <img>, <audio> and <video> elements should use dataURIs
    bundle: bool,
}

impl<'a> EncodeContext<'a> {
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        EncodeContext {
            root: &Node::Null(Null {}),
            bundle: false,
        }
    }
}

/// Trait for encoding a node as HTML
pub trait ToHtml {
    fn to_html(&self, context: &EncodeContext) -> String;
}

/// Create an empty string
///
/// Just used to make it clear that not encoding anything to HTML (usually
/// because a property is missing).
fn nothing() -> String {
    String::new()
}

/// Encode a HTML element
///
/// Use this function for creating HTML strings for elements.
/// This, and other functions below, us slice `concat`, rather than `format!`
/// for performance (given that HTML generation may be done on every, or nearly every, keystroke).
fn elem(name: &str, attrs: &[String], content: &str) -> String {
    let attrs = attrs.iter().fold(String::new(), |mut attrs, attr| {
        if !attr.is_empty() {
            if !attrs.is_empty() {
                attrs.push(' ');
            }
            attrs.push_str(attr);
        }
        attrs
    });
    [
        "<",
        name,
        if attrs.is_empty() { "" } else { " " },
        &attrs,
        ">",
        content,
        "</",
        name,
        ">",
    ]
    .concat()
}

/// Encode an "empty" HTML element
///
/// An empty (a.k.a self-closing) element has no closing tag.
/// See https://developer.mozilla.org/en-US/docs/Glossary/Empty_element
fn elem_empty(name: &str, attrs: &[String]) -> String {
    [
        "<",
        name,
        if attrs.is_empty() { "" } else { " " },
        attrs.join(" ").trim(),
        "/>",
    ]
    .concat()
}

/// Encode a `<meta>` attribute
///
/// This is used to encode simple, usually string, properties of nodes as HTML Microdata
/// when the property should not be visible, or is represented some other way.
fn elem_meta(name: &str, content: &str) -> String {
    elem_empty("meta", &[attr_itemprop(name), attr("content", content)])
}

/// Encode an optional property as an element
///
/// If the property is `None` then the element will be empty but will act
/// as a placeholder for if/when the property is set.
/// By having placeholders the order of optional properties in the HTML tree
/// can be consistent when they are added and removed.
fn elem_placeholder<T: ToHtml>(
    name: &str,
    attrs: &[String],
    content: &Option<T>,
    context: &EncodeContext,
) -> String {
    elem(
        name,
        attrs,
        &match content {
            Some(content) => content.to_html(context),
            None => nothing(),
        },
    )
}

/// Encode a HTML element attribute, ensuring that the value is escaped correctly
fn attr(name: &str, value: &str) -> String {
    [
        name,
        "=\"",
        encode_double_quoted_attribute(&value).as_ref(),
        "\"",
    ]
    .concat()
}

/// Encode one of the attributes used to identify a property of a Stencila node
///
/// Will ensure that the name is camelCased.
fn attr_prop(name: &str) -> String {
    if name.is_empty() {
        "".to_string()
    } else {
        attr("data-itemprop", &to_camel_case(name))
    }
}

/// Encode the "itemtype" attribute of an HTML element
///
/// Prefer to use `attr_itemtype::<Type>` but use this when the
/// itemtype should differ from the Rust type name.
/// Note: there should always be a sibling "itemscope" attribute on the
/// element so that is added.
fn attr_itemtype_str(name: &str) -> String {
    let itemtype = match name {
        // TODO: complete list of schema.org types
        "Article" | "AudioObject" | "ImageObject" | "VideoObject" | "Text" => {
            ["http://schema.org/", name].concat()
        }
        _ => ["http://schema.stenci.la/", name].concat(),
    };
    [&attr("itemtype", &itemtype), " itemscope"].concat()
}

/// Encode the "itemtype" attribute of an HTML element using the type of node
fn attr_itemtype<Type>() -> String {
    let name = type_name::<Type>();
    let name = if let Some(name) = name.strip_prefix("stencila_schema::types::") {
        name
    } else {
        tracing::error!("Unexpected type: {}", name);
        name
    };
    attr_itemtype_str(name)
}

/// Encode a "itemprop" attribute of an HTML element
///
/// Will ensure that the itemprop is camelCased.
fn attr_itemprop(itemprop: &str) -> String {
    if itemprop.is_empty() {
        "".to_string()
    } else {
        attr("itemprop", &to_camel_case(itemprop))
    }
}

/// Encode a node `id` as the "id" attribute of an HTML element
fn attr_id(id: &Option<Box<String>>) -> String {
    match id.as_deref() {
        Some(id) => attr("id", id),
        None => "".to_string(),
    }
}

/// Encode the "slot" attribute of an HTML
///
/// Used for nodes that are represented in HTML using a custom Web Component.
/// Not to be confused with the Stencila `Address` slot which will often have the
/// same value but which will be encoded as a "data-prop" if the element is not a Web Component.
fn attr_slot(name: &str) -> String {
    attr("slot", name)
}

/// Encode a node as JSON
///
/// Several of the below implementations use this, mainly as a placeholder,
/// until a complete implementation is finished. Ensures that the JSON is
/// properly escaped
fn json(node: &impl serde::Serialize) -> String {
    encode_safe(&serde_json::to_string_pretty(node).unwrap_or_default()).to_string()
}

/// Iterate over a slice of nodes, call a string generating function on each item
/// and concatenate the strings
pub fn concat<T, F>(slice: &[T], func: F) -> String
where
    F: FnMut(&T) -> String,
{
    slice.iter().map(func).collect::<Vec<String>>().concat()
}

mod blocks;
mod boxes;
mod inlines;
mod nodes;
mod options;
mod primitives;
mod vecs;
#[allow(clippy::deprecated_cfg_attr)]
mod works;

#[cfg(test)]
mod tests {
    use super::*;
    use crate::decode::decode;
    use codec::eyre::bail;
    use serde_json::json;
    use test_snaps::{
        insta::assert_display_snapshot, snapshot_fixtures_content, snapshot_fixtures_nodes,
    };
    use test_utils::{assert_json_eq, home, skip_slow_tests};

    /// Encode the node fixtures
    #[test]
    fn encode_nodes() {
        snapshot_fixtures_nodes("nodes/*.json", |node| {
            let html = encode(
                &node,
                Some(EncodeOptions {
                    compact: false,
                    ..Default::default()
                }),
            )
            .unwrap();
            assert_display_snapshot!(html);
        });
    }

    /// Encode the HTML fragment fixtures (involves decoding them first)
    #[test]
    fn encode_html_fragments() {
        snapshot_fixtures_content("fragments/html/*.html", |content| {
            let decoded = decode(content, None).expect("Unable to decode");
            let encoded = encode(
                &decoded,
                Some(EncodeOptions {
                    compact: false,
                    ..Default::default()
                }),
            )
            .unwrap();
            assert_display_snapshot!(encoded);
        });
    }

    /// Validate HTML against https://validator.github.io/validator/
    ///
    /// To run locally using the validator's Docker image:
    ///
    ///  docker run -it --rm -p 8888:8888 ghcr.io/validator/validator
    ///  RUN_SLOW_TESTS=1 HTML_VALIDATOR=http://localhost:8888 cargo test
    ///
    /// See https://github.com/validator/validator/wiki/Service-%C2%BB-Input-%C2%BB-POST-body
    /// for more on the API.
    #[tokio::test]
    async fn nu_validate() -> Result<()> {
        if skip_slow_tests() {
            return Ok(());
        }

        // Read the existing snapshot
        // We only do this for one, kitchen sink like, snapshot.
        let html = std::fs::read_to_string(
            home().join("rust/codec-html/src/snapshots/encode_html_fragments@heading.html.snap"),
        )?;
        let decoded = decode(&html, None)?;
        let html = encode(
            &decoded,
            Some(EncodeOptions {
                standalone: true,
                compact: false,
                ..Default::default()
            }),
        )?;

        // Make the POST request
        let url = if let Ok(url) = std::env::var("HTML_VALIDATOR") {
            url
        } else {
            "https://validator.w3.org/nu/".to_string()
        };
        let client = reqwest::Client::new();
        let response = client
            .post([&url, "?out=json"].concat())
            .header("Content-Type", "text/html; charset=UTF-8")
            .header("Accept", "application/json")
            .header(
                "User-Agent",
                "Stencila tests (https://github.com/stencila/stencila/)",
            )
            .body(html)
            .send()
            .await?;
        let response = match response.error_for_status() {
            Ok(response) => response,
            Err(error) => bail!(error),
        };
        let json = response.text().await?;

        // Parse the result so it's easier to read any messages
        let result: serde_json::Value = serde_json::from_str(&json)?;
        assert_json_eq!(result, json!({"messages": []}));

        Ok(())
    }
}
