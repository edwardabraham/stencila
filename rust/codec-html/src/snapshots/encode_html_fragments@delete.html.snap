---
source: rust/codec-html/src/encode.rs
assertion_line: 442
expression: encoded
input_file: fixtures/fragments/html/delete.html

---
<article data-itemscope="root" itemtype="http://schema.org/Article" itemscope>
  <stencila-document-toolbar>
  </stencila-document-toolbar>
  <div data-itemprop="content">
    <p itemtype="http://schema.stenci.la/Paragraph" itemscope>
      <span>Some</span>
      <del itemtype="http://schema.stenci.la/Delete" itemscope>
        <span>deleted</span>
      </del>
      <span>text.</span>
    </p>
  </div>
</article>
